ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java

```diff
@@ -275,20 +294,22 @@ private Response executeStatements(
     return Response.ok(entities).build();
   }
 
+  @SuppressWarnings("unchecked")
   private <T extends Statement> KsqlEntity executeStatement(
       final PreparedStatement<T> statement,
       final Map<String, Object> propertyOverrides,
       final KsqlEntityList entities
   ) {
     try {
-      final Handler<T> handler = getCustomExecutor(statement);
-      if (handler != null) {
-        waitForPreviousDistributedStatementToBeHandled(entities);
+      final Class<? extends Statement> type = statement.getStatement().getClass();
+      final Handler<T> handler =
```

Comment:
You've changed the logic here of when a statement waits for a previous statement to execute.

The old code does not wait for anything that doesn't have a custom handler. So, by default, if the statement is using the default path and being distributed to the command topic, it does _NOT_ wait on previous commands.

This is by design, as commands that are distributed are synchronised with earlier commands by the command topic. Only commands that don't use the command topic _may_ need explicit synchronisation.

With your change all the commands that are distributed _ARE_ waiting for previous commands to execute.

There is actually a test to check that distributed commands do not wait for previous ones: `shouldNotWaitOnAnyDistributedStatementsBeforeDistributingAnother`.  Maybe this test needs more than one statement type in it!

Reply:
Before your explanation I didn't understand why we were tying to seemingly distinct state together (having a custom handler and needing to wait before executing). With this explanation, it makes a lot more sense. I think maybe we can rename some things to make it clearer for first-time readers or add some in-line comments. I'll take a stab at that.